### PR TITLE
Changed to 24h time format

### PR DESCRIPTION
**Pull request by smitbose**
This is a fix for issue #675 

This changes the time shown in the chatbox from a 12h time format to a 24h time format.

### CODE REVIEW

**Comment by smitbose** [comment]
screenshot from 2017-03-04 02-19-58

**Comment by jcbrand** [comment]
The point of that ticket is to make the format configurable, not to hardcode it to a new format.

So you want to add a new configuration setting, something like `date_format`, let it have the original format as it's default and then let the user specify a different format if they want one.

**Comment by smitbose** [comment]
ok, @jcbrand I get it now, I'll try to generate a new patch with the required change.

**Comment by smitbose** [comment]
@jcbrand I have added a setting time_format, that takes vales 'h' and 'H' for 12 and 24 hr time formats respectively. Also, for the 12hr time format, it will display am or pm, because personally I found this am or pm not being displayed as making reading of previous messages too ambiguous.
screenshot from 2017-03-08 11-14-26

File: src/converse-chatview.js [hunk 1]
```diff
@@ -344,11 +344,19 @@
...
+                        mess_time = msg_time.format('hh:mm a');
+                    else if(_converse.time_format === 'H')
+                        mess_time = msg_time.format('HH:MM');
+
                     var $msg = $(template(
                         _.extend(this.getExtraMessageTemplateAttributes(attrs), {
                             'msgid': attrs.msgid,
                             'sender': attrs.sender,
-                            'time': msg_time.format('hh:mm'),
+                            'time': mess_time,
```
Comment (smitbose) [inline src/converse-chatview.js]: @jcbrand in one of the above comments, you have told to keep the original one(which according to current release is hh:mm) as the default. Can you confirm once again, which one will I finally keep .

Moreover, doing the above check helps in printing am or pm beside the time in 12hr format. Because for that moment needs hh:mm a. One way is of-course to let the user specify 'hh:mm a' as the setting. And the other way would be to simply leave the printing of 'am' or 'pm' in messages, which I feel makes a message sent at 6am and 6pm confusing.

Comment (jcbrand) [inline src/converse-chatview.js]: My original point had less to do with what the default was and more to do with the fact that you hardcoded the change.

I think I prefert 24-hour time as well, so I'd prefer HH:MM.

Concerning showing AM and PM, since we want to make the user (really the integrator) responsible for choosing the format, they can decide whether to show AM/PM or not.

Comment (smitbose) [inline src/converse-chatview.js]: ok, so I'm making a change with, keeping three settings values then, 'HH:MM', 'hh:mm' and 'hh:mm a', which will give a 24hr time format, 12hr time format, and a 12hr time format with am/pm displayed.

Comment (jcbrand) [inline src/converse-chatview.js]: No. That's not what I meant. We want the user to specify the format. Whatever format that is. Don't try to restrict the format that the user can provide.

File: src/converse-core.js [hunk 2]
```diff
@@ -280,6 +281,10 @@
             }
         }
 
+        if(this.time_format !== 'h' && this.time_format !== 'H') {
+            throw new Error("Config Error: wrong time format configuration");
+        }
+
```
Comment (jcbrand) [inline src/converse-core.js]: This must be removed since we don't want to restrict the values the user can enter as the format.

File: src/converse-core.js [hunk 3]
```diff
@@ -267,7 +267,7 @@
             whitelisted_plugins: [],
             xhr_custom_status: false,
             xhr_custom_status_url: '',
-            time_format: 'h', // H stands for 24hr time format and 'h' stands for 12hr time format
+            time_format: 'HH:MM', // H stands for 24hr time format and 'h' stands for 12hr time format
```
Comment (jcbrand) [inline src/converse-core.js]: Please remove the comment.

File: src/converse-core.js [hunk 4]
```diff
@@ -281,7 +281,7 @@
             }
         }
 
-        if(this.time_format !== 'h' && this.time_format !== 'H') {
+        if(this.time_format !== 'hh:mm' && this.time_format !== 'HH:MM' && this.time_format !== 'hh:mm a') {
```
Comment (jcbrand) [inline src/converse-core.js]: Remove this whole `if` statement. We don't want to restrict what format the user can provide.

Comment (smitbose) [inline src/converse-core.js]: okay.

File: src/converse-chatview.js [hunk 5]
```diff
@@ -344,11 +344,13 @@
                                "Output has been shortened."),
                             true, true);
                     }
+
+
```
Comment (jcbrand) [inline src/converse-chatview.js]: Unnecessary whitespace.

File: src/converse-core.js [hunk 6]
```diff
@@ -267,6 +267,7 @@
             whitelisted_plugins: [],
             xhr_custom_status: false,
             xhr_custom_status_url: '',
+            time_format: 'HH:MM', // H stands for 24hr time format and 'h' stands for 12hr time format
```
Comment (jcbrand) [inline src/converse-core.js]: Please remove the comment.

**Comment by jcbrand** [comment]
@smitbose Almost ready for merge. What still needs to happen is the documentation needs to be updated.

Please add an entry for `time_format` in the configuration settings page:
https://github.com/jcbrand/converse.js/blob/master/docs/source/configuration.rst

Some info on writing documentation: https://conversejs.org/docs/html/documentation.html

Then also update the changelog in `docs/CHANGES.md`

File: docs/source/configuration.rst [hunk 7]
```diff
@@ -1067,6 +1067,17 @@ If set to ``false``, this feature is disabled.
 
 If set to ``a resource name``, converse.js will synchronize only with a client that has that particular resource assigned to it.
 
+time_format
+-----------
+
+* Default: ``HH:MM``
+
+Valid option: ``HH:MM``, ``hh:mm``, ``hh:mm a``.
+
```
Comment (smitbose) [inline docs/source/configuration.rst]: okay

File: docs/source/configuration.rst [hunk 8]
```diff
@@ -1067,6 +1067,17 @@ If set to ``false``, this feature is disabled.
...
+This option makes the time format for the time shown, for each message, configurable. Converse can be configured to show 
+the time in 24hr or 12hr format. The default value gives a 24hr time-format. To change it to 12hr, use `hh:mm`, and to display
+AM/PM, use `hh:mm a`.
+
```
Comment (jcbrand) [inline docs/source/configuration.rst]: Also the text here is too specific, it makes me think that you're still missing the point somewhat, I think because of the way the original ticket was specified.

What we are doing is letting the user specify **any** format for message time, they can even include the date if they want to. So don't say it's either 24hr or 12hr, just say that it can show any format that moment.js supports.

Comment (smitbose) [inline docs/source/configuration.rst]: okay, I'll try to make it more clear

**Comment by jcbrand** [comment]
One more thing. Please add a test to show that different formats can be used. You can add this test in [spec/chatbox.js](https://github.com/jcbrand/converse.js/blob/master/spec/chatbox.js).

Let me know if you need help.

**Comment by smitbose** [comment]
@jcbrand as far as the documentation is concerned, adding a link to moment.js will give the user a better idea of the different formats they can use, along with an explanation with an instance. 
As for the tests, would just checking for the value of `_converse.time_format` suffice, or should I check the time of a dummy message sent?

**Comment by jcbrand** [comment]
+1 for adding a link (if you find a place in the moment.js docs where the different formats are shown).

Concerning the test: set a non-default value for the `time_format` and the check that a sent messages renders correctly.

**Comment by smitbose** [comment]
https://momentjs.com/docs/#/parsing/string-format/

I think this link makes a place in the doc. What say?

File: spec/chatbox.js [hunk 9]
```diff
@@ -1240,6 +1240,33 @@
                     });
                 }));
 
+                it("will render the message time as configured", mock.initConverse(function (_converse) {
+                    test_utils.createContacts(_converse, 'current');
+                    test_utils.openControlBox();
+                    test_utils.openContactsPanel(_converse);
+
```
Comment (jcbrand) [inline spec/chatbox.js]: I think lines 1245 and 1246 can be removed.

File: spec/chatbox.js [hunk 10]
```diff
@@ -1240,6 +1240,33 @@
...
+                    _converse.time_format = 'hh:mm';
+                    spyOn(_converse, 'emit');
+                    var contact_jid = mock.cur_names[0].replace(/ /g,'.').toLowerCase() + '@localhost';
+                    test_utils.openChatBoxFor(_converse, contact_jid);
+                    expect(_converse.emit).toHaveBeenCalledWith('chatBoxFocused', jasmine.any(Object));
```
Comment (jcbrand) [inline spec/chatbox.js]: Irrelevant check which can be removed. You can then also remove the `spyOn(_converse, 'emit');` line.

File: spec/chatbox.js [hunk 11]
```diff
@@ -1240,6 +1240,33 @@
...
+                    var view = _converse.chatboxviews.get(contact_jid);
+                    var message = 'This message is sent from this chatbox';
+                    spyOn(view, 'sendMessage').andCallThrough();
+                    test_utils.sendMessage(view, message);
+                    expect(view.sendMessage).toHaveBeenCalled();
+                    expect(view.model.messages.length, 2);
```
Comment (jcbrand) [inline spec/chatbox.js]: The two `expect` lines above are irrelevant for this test and should be removed. Also the `spyOn` above that.

File: spec/chatbox.js [hunk 12]
```diff
@@ -1240,6 +1240,33 @@
...
+
+                    var chatbox = _converse.chatboxes.get(contact_jid);
+                    expect(chatbox.messages.models.length, 1);
+                    var msg_object = chatbox.messages.models[0];
+                    var msg_time_author = view.$el.find('.chat-content').find('.chat-message').last().find('.chat-msg-author.chat-msg-me').text();
```
Comment (jcbrand) [inline spec/chatbox.js]: Wrap this line please, so that it's not so long.

File: docs/CHANGES.md [hunk 13]
```diff
@@ -4,6 +4,7 @@
 
 - #585 Duplicate contact created due to JID case sensivity [saganshul]
 - #628 Fixes the bug in displaying chat status during private chat. [saganshul]
+- #797 Time format made configurable, with a choice between 12hr and 24hr format. [smitbose]
```
Comment (jcbrand) [inline docs/CHANGES.md]: Update this entry please. Still falsely creates the impression that there's 2 choices.

**Comment by jcbrand** [comment]
@smitbose I'm not sure whether you were still busy with this or already waiting for review, but I went ahead and did a quick review anyway.

File: docs/CHANGES.md (hunk 13 above)
Comment (smitbose) [inline docs/CHANGES.md]: yeah @jcbrand I think this 12hr/24hr thing has got embedded in my head! :P I keep repeating the same damn thing every single time! :D

**Comment by smitbose** [comment]
@jcbrand please review

**Comment by jcbrand** [comment]
Looking good thanks!
